src/main/java/org/openrewrite/java/template/RefasterTemplateProcessor.java

```diff
@@ -187,24 +187,34 @@ public void visitClassDef(JCTree.JCClassDecl classDecl) {
                     for (JCTree.JCMethodDecl template : descriptor.beforeTemplates) {
                         for (Symbol anImport : ImportDetector.imports(template)) {
                             if (anImport instanceof Symbol.ClassSymbol) {
-                                imports.add(anImport.getQualifiedName().toString().replace('$', '.'));
+                                imports.computeIfAbsent(template, k -> new TreeSet<>())
+                                        .add(anImport.getQualifiedName().toString().replace('$', '.'));
                             } else if (anImport instanceof Symbol.VarSymbol || anImport instanceof Symbol.MethodSymbol) {
-                                staticImports.add(anImport.owner.getQualifiedName().toString().replace('$', '.') + '.' + anImport.flatName().toString());
+                                staticImports.computeIfAbsent(template, k -> new TreeSet<>())
+                                        .add(anImport.owner.getQualifiedName().toString().replace('$', '.') + '.' + anImport.flatName().toString());
                             } else {
                                 throw new AssertionError(anImport.getClass());
                             }
                         }
                     }
                     for (Symbol anImport : ImportDetector.imports(descriptor.afterTemplate)) {
                         if (anImport instanceof Symbol.ClassSymbol) {
-                            imports.add(anImport.getQualifiedName().toString().replace('$', '.'));
+                            imports.computeIfAbsent(descriptor.afterTemplate, k -> new TreeSet<>())
+                                    .add(anImport.getQualifiedName().toString().replace('$', '.'));
                         } else if (anImport instanceof Symbol.VarSymbol || anImport instanceof Symbol.MethodSymbol) {
-                            staticImports.add(anImport.owner.getQualifiedName().toString().replace('$', '.') + '.' + anImport.flatName().toString());
+                            staticImports.computeIfAbsent(descriptor.afterTemplate, k -> new TreeSet<>())
+                                    .add(anImport.owner.getQualifiedName().toString().replace('$', '.') + '.' + anImport.flatName().toString());
                         } else {
                             throw new AssertionError(anImport.getClass());
                         }
                     }
 
+                    for (Set<String> imports : imports.values()) {
+                        imports.removeIf(i -> "java.lang".equals(i.substring(0, i.lastIndexOf('.'))));
```

Comment:
This is better indeed; the exclude of `java.lang.**` we had before was too broad. Thanks!